### DISCUSSION TITLE
how to avoid typescript errors in Attributify mode?

### DISCUSSION THREAD

**Discussion by MarkLyck** [unocss/unocss#740] [Q&A]
After setting up unocss like this in my vite.config.ts:

```
import path from 'path'
import { defineConfig } from 'vite'

// plugins
import ssr from 'vite-plugin-ssr/plugin'
import react from '@vitejs/plugin-react'
import Unocss from 'unocss/vite'
import { presetAttributify, presetUno } from 'unocss'

export default defineConfig({
  resolve: {
    alias: {
      'src/': `${path.resolve(__dirname, 'src')}/`,
    },
  },
  plugins: [
    react(),
    ssr(),
    Unocss({
      presets: [presetAttributify(), presetUno()],
    }),
  ],
})
```

and trying a super simple example component:

```
import React from 'react'

const Test = () => {
  return (
        <p text="red-500">test</p>
  )
}

export default Test
```

I get this typescript  error:

```
pages/index/Hero/index.tsx:8:12 - error TS2322: Type '{ children: string; text: string; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLParagraphElement>, HTMLParagraphElement>'.
  Property 'text' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLParagraphElement>, HTMLParagraphElement>'.

8         <p text="red-500">test</p>
```

![Screen Shot 2022-03-19 at 12 54 07](https://user-images.githubusercontent.com/6841110/159130624-1ab22c93-104e-4305-9971-a086ac7729c3.png)

Which makes sense in "normal" jsx. But with attributify this should be valid?

The text does work and it does become a red color, so the mode itself works. But my typescript doesn't like it?

This is my tsconfig.json file setup:

```
{
  "compilerOptions": {
    "strict": true,
    "target": "ESNext",
    "module": "esnext",
    "moduleResolution": "Node",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "types": ["vite/client"],
    "jsx": "react",
    "skipLibCheck": true,
    "esModuleInterop": true
  },
  "ts-node": {
    "transpileOnly": true,
    "compilerOptions": {
      "module": "CommonJS"
    }
  }
}
```

**Comment by KKandLL-Forever** [comment]
I had same problem before.
try https://github.com/unocss/unocss/blob/main/packages/preset-attributify/README.md#typescript-support-jsxtsx

**Comment by sjmueller** [comment]
For those arriving here via google search `vscode Argument of type '{ text: string; }' is not assignable to parameter of type 'HTMLProps<"h1", HTMLAttributes<any>>'`

I am sharing more comprehensive instructions, since the link @KKandLL-Forever provided no longer has the section `typescript-support-jsxtsx`

The desired documentation is at https://unocss.dev/presets/attributify#typescript-support-jsx-tsx ***however*** it is currently incomplete and missing the necessary code for `shims.d.ts`. The correct code is in the nextjs example [found here](https://github.com/unocss/unocss/blob/main/examples/next/shims.d.ts):

``` ts
// shims.d.ts
// this file is only needed if you use `@unocss/preset-attributify`
import type { AttributifyAttributes } from '@unocss/preset-attributify'

declare module 'react' {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  interface HTMLAttributes<T> extends AttributifyAttributes { }
}
```

### `sveltekit`
The above code won't work for sveltekit. Instead, modify `./src/app.d.ts`:

``` ts
// app.d.ts
import type { AttributifyAttributes } from "@unocss/preset-attributify"

// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
  namespace App {
    // interface Error {}
  }

  namespace svelte.JSX {
    interface HTMLAttributes<T> extends AttributifyAttributes { }
  }
}

export {}
``` 

I will submit a PR that fixes the documentation and adds the appropriate code to `examples/sveltekit-scoped`.
